### ISSUE TITLE
ttnn.remainder unary low PCC when scalar is between -0.003 and 0.003

### ISSUE THREAD

**Issue by amalbasaTT** [tenstorrent/tt-metal#17361]
**To Reproduce**
Run unit test bellow:
```
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback

from tests.ttnn.utils_for_testing import check_with_pcc
from tests.sweep_framework.sweep_utils.utils import tensor_to_dtype
from models.utility_functions import skip_for_grayskull


def run_remainder_unary_tests(
    input_shape,
    scalar,
    dtype,
    dlayout,
    in_mem_cfg,
    out_mem_cfg,
    device,
):
    torch.manual_seed(0)
    
    x = tensor_to_dtype(torch.Tensor(size=input_shape).uniform_(-100, 100), dtype)

    try:
        # get ref result
        golden_function = ttnn.get_golden_function(ttnn.remainder)
        ref_value = golden_function(x, scalar)
        
        tt_input = ttnn.from_torch(x, dtype=dtype, layout=dlayout, device=device, memory_config=in_mem_cfg)

        tt_result = ttnn.remainder(tt_input, scalar, memory_config=out_mem_cfg)
        tt_result = ttnn.to_torch(tt_result)

    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    pcc, message = check_with_pcc(ref_value, tt_result, 0.999)
    assert pcc, message
    

test_sweep_args = [
    (
        (6, 5, 90, 112),
        torch.tensor(1, dtype=torch.bfloat16).uniform_(-0.003, 0.003).item(),
        ttnn.bfloat16,
        ttnn.TILE_LAYOUT,
        ttnn.DRAM_MEMORY_CONFIG,
        ttnn.DRAM_MEMORY_CONFIG,
    ),
    (
        (2, 1, 120, 11),
        torch.tensor(1, dtype=torch.bfloat16).uniform_(-0.003, 0.003).item(),
        ttnn.bfloat16,
        ttnn.TILE_LAYOUT,
        ttnn.L1_MEMORY_CONFIG,
        ttnn.DRAM_MEMORY_CONFIG,
    ),
    (
        (3, 123, 115),
        torch.tensor(1, dtype=torch.bfloat16).uniform_(-0.003, 0.003).item(),
        ttnn.bfloat16,
        ttnn.TILE_LAYOUT,
        ttnn.DRAM_MEMORY_CONFIG,
        ttnn.L1_MEMORY_CONFIG,
    ),
    (
        (69, 178),
        torch.tensor(1, dtype=torch.bfloat16).uniform_(-0.003, 0.003).item(),
        ttnn.bfloat16,
        ttnn.TILE_LAYOUT,
        ttnn.L1_MEMORY_CONFIG,
        ttnn.L1_MEMORY_CONFIG,
    ),
]



@pytest.mark.parametrize(
    "input_shape, scalar, dtype, dlayout, in_mem_config, out_mem_config",
    (test_sweep_args),
)
@skip_for_grayskull("Not supported for Grayskull")
def test_remainder_unary(input_shape, scalar, dtype, dlayout, in_mem_config, out_mem_config, device):
    run_remainder_unary_tests(input_shape, scalar, dtype, dlayout, in_mem_config, out_mem_config, device)

```

**Comment by mouliraj-mcw** [comment]
The problem stems from the use of float_to_int16 conversion in the kernel. I've detailed the issue further in the document linked here. [Link](https://docs.google.com/document/d/1Z0wn60JIwHXrR8t5Eqnm0wj17N5GWvfgmC-cQ4AV6RA/edit?usp=sharing).
Hi @rdjogoTT . Please have a look.

**Comment by rdjogoTT** [comment]
@mouliraj-mcw since the goal of float->int->float is to perform floor, you can use a kernel like this to accomplish that (will floor for positive numbers by cutting of decimal component to 0). Maybe this will help with the register spill issue.

```
vFloat in = dst_reg[0];

vInt exp = exexp(in);
v_if (exp < 0) {
    dst_reg[0] = vConst0;
} v_elseif (exp < 23) {
    vInt shift = exp - 23;
    vUInt temp = shft(reinterpret<vUInt>(in), shift);
    vFloat result = reinterpret<vFloat>(shft(temp, (0 - shift)));
    dst_reg[0] = result;
} v_else {
    dst_reg[0] = in;
} v_endif
```

Regarding bad PCC due to recip and quotient inaccuracy, there isn't anything that can be done if the values remain in fp16b. What you can try is to load `value` and `recip` as fp32 to help with accuracy. One was this should be possible is to load them into the programmable const regs in the init function, something like this should work:

```
#include "ckernel_sfpu_load_config.h"

template <bool APPROXIMATION_MODE>
inline void _init_remainder_(const uint value, const uint recip)
{
    // load vConstFloatPrgm0 = value
    _sfpu_load_config32_(0xC, (value >> 16) & 0xFFFF, value & 0xFFFF);
    // load vConstFloatPrgm1 = recip
    _sfpu_load_config32_(0xD, (recip >> 16) & 0xFFFF, recip & 0xFFFF);
}
```
You can then use vConstFloatPrgm0 and vConstFloatPrgm1 to get value and recip.

Give these a try and let me know how it goes.
If these suggestions do not help with PCC, a different algorithm for remainder might be needed.

**Comment by mouliraj-mcw** [comment]
@rdjogoTT, I'll try those out and update you soon.

**Comment by mcw-anasuya** [comment]
@rdjogoTT I tried to load value and recip as fp32 in this [PR](https://github.com/tenstorrent/tt-metal/pull/19008) as you suggested, however it did not affect the output. I tested with a small input matrix with scalar = 0.003 and checked the atol value. Looks like the outputs from the existing logic in main and the logic in the above mentioned PR are exactly the same.
<img width="730" alt="Image" src="https://github.com/user-attachments/assets/129a48a0-d222-427c-8473-9aeae181ec60" />

I also tested the same matrix with scalar = 0.004 and the test passes even though the atol value is larger.
<img width="735" alt="Image" src="https://github.com/user-attachments/assets/803da5f7-704d-45ea-b219-9b3544b3fcd0" />

**Comment by rdjogoTT** [comment]
@mcw-anasuya so looking at the scalar=0.004 case you have there, the torch output seems to be wrong as well. It's returning 0.004 but the remainder should just be 0. Also, if all of the values in one output (either torch or ttnn) are constants, then the PCC calculation gets messed up. So this explains why the test is passing despite being less accurate than the other test.

I think you need to re-evaluate how accuracy is measured for this OP, since PCC seems not to be the best and also torch outputs seem strange.
I can't look into why the accuracy is low now, but you can file an issue under LLK tag and someone from the team will help when they can.

**Comment by mcw-anasuya** [comment]
@amalbasaTT Few observations regarding ttnn.remainder and ttnn.fmod (#17362):
1. The overall accuracy of the ops has improved with the modifications [here](https://github.com/tenstorrent/tt-metal/issues/17361#issuecomment-2691371528).
2. The torch outputs for `bfloat16` was not as accurate as the TT outputs. I updated the golden function to convert `bf16` to `fp32` before the torch functions are run.
3. There is ambiguity in both Torch and TT outputs when the scalar value is small and it is a factor of the input tensor value i.e. both Torch and TT do not output 0.0 in all cases where the input is a multiple of the scalar value.
**Ambiguity in Torch:**
`torch.remainder(torch.tensor([3]), 1.5) = 0.0`
`torch.remainder(torch.tensor([3]), 0.003) = 0.003`
**Ambiguity in TT:**
`ttnn.remainder([3], 0.003) = 0.0`
`ttnn.remainder([4], 0.004) = 0.004`
This inconsistency is caused by a precision loss that occurs while storing the data (the scalar `0.003` may be stored as `0.0030001`). Since the inconsistency is persistent in both Torch and TT, it causes the PCC to be low for some scalars in the range `[-0.003, 0.003]`. As a result, I've used torch.allclose as a better comparison metric (it would still fail for some scalar values like `0.003` since the atol value is set at `0.001`).

**Comment by mcw-anasuya** [comment]
The PR increasing the overall accuracy has been merged to main. Hence closing this issue.
